File: C++/Easy/125. Valid Palindrome.cpp
// https://leetcode.com/problems/valid-palindrome/description/

class Solution {
public:
    bool isPalindrome(string s) {
        if(s.length() < 2) {
            return true;
        }
        s.erase(std::remove_if(s.begin(), s.end(), [](unsigned char c) { return !std::isalnum(c); }), s.end());
        transform(s.begin(), s.end(), s.begin(), [](unsigned char c){ return std::tolower(c); });
        int n = s.length();

        for(int i = 0; i < n/2; i++) {
            if (s[i] != s[n - 1 - i]) {
                return false;
            }
        }
        return true;
    }
};

File: Java/Easy/125. Valid Palindrome.java
// https://leetcode.com/problems/valid-palindrome/description/

class Solution {
    public boolean isPalindrome(String s) {
        s = s.toLowerCase(); 
        StringBuilder filtered = new StringBuilder();

        for (char ch : s.toCharArray()) {
            if (Character.isLetterOrDigit(ch)) {
                filtered.append(ch);
            }
        }
        
        int n = filtered.length();
        for (int i = 0; i < n / 2; i++) {
            if (filtered.charAt(i) != filtered.charAt(n - 1 - i)) {
                return false;
            }
        }
        return true;
    }
}